shared-module/audiofilters/Filter.c

```diff
@@ -87,15 +87,37 @@ void common_hal_audiofilters_filter_deinit(audiofilters_filter_obj_t *self) {
     self->buffer[0] = NULL;
     self->buffer[1] = NULL;
     self->filter_buffer = NULL;
+    self->filter_states = NULL;
 }
 
-mp_obj_t common_hal_audiofilters_filter_get_filter(audiofilters_filter_obj_t *self) {
-    return self->filter_obj;
+void reset_filter_states(audiofilters_filter_obj_t *self) {
+    self->filter_states_len = self->filters->len;
+    self->filter_states = NULL;
+
+    if (self->filter_states_len) {
+        self->filter_states = m_malloc(self->filter_states_len * sizeof(biquad_filter_state));
+        if (self->filter_states == NULL) {
+            common_hal_audiofilters_filter_deinit(self);
+            m_malloc_fail(self->filter_states_len * sizeof(biquad_filter_state));
```

Comment:
I'm relatively sure that `m_malloc` already throws in case of failed allocation, so the call to `m_malloc_fail` is not needed; but also, the call to `common_hal_audiofilters_filter_deinit` is ineffective. I see this idiom has been used across the audiofilters so it may need to be revised elsewhere as well.

Reply:
I'll look into this issue and apply changes as necessary throughout. I'm still learning about how best to program within the MicroPython/CircuitPython ecosystem so bear with me! :)

Reply:
I began removing the `..._deinit` and `m_malloc_fail` statements, but I'm finding that this idiom is fairly common within multiple modules of CircuitPython. As you certain that this isn't the proper handling of `m_malloc`?

Here are a few examples:
- https://github.com/adafruit/circuitpython/blob/ddfa5193ec41212c1c4db4015d9f6298a973d5b3/shared-module/audiomixer/Mixer.c#L28-L38
- https://github.com/adafruit/circuitpython/blob/ddfa5193ec41212c1c4db4015d9f6298a973d5b3/shared-module/audiomp3/MP3Decoder.c#L306-L327
- https://github.com/adafruit/circuitpython/blob/ddfa5193ec41212c1c4db4015d9f6298a973d5b3/shared-module/audiocore/WaveFile.c#L100-L110
- https://github.com/adafruit/circuitpython/blob/ddfa5193ec41212c1c4db4015d9f6298a973d5b3/ports/stm/common-hal/pulseio/PulseIn.c#L87-L91 (also in other ports with pulsio)

There are other cases where instead of `m_malloc(...)` there are calls to `port_malloc(...)`, `malloc(...)`, `realloc(...)`, `malloc_with_finaliser(...)`, `ringbuf_alloc(...)`, and `gc_alloc(...)` before `m_malloc_fail(...)`. Maybe someone down the line copied that over but with `m_malloc(...)` and it stuck within audio and pulseio modules?

Rather than experimenting and going outside of the accepted formatting, I'm copying a lot of my homework from other portions of CircuitPython. This is something I borrowed from `audiodelays` without fully understanding the underlying principles. I plan on keeping it as is for now until we develop a plan for all other instances of this methodology (likely another PR).

Reply:
Finally looked this up, and I see what you mean. The additional check and `m_malloc_fail` is definitely not necessary. I'll remove it from this class, but let me know if you'd like me to go around and remove it throughout CircuitPython or save it for a separate PR. @jepler 

https://github.com/adafruit/circuitpython/blob/d7a72218159f9e88e801949a74d530225a50320a/py/malloc.c#L86-L89